src/runtime-rs/crates/resource/src/rootfs/block_rootfs.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use super::{Rootfs, ROOTFS};
+use crate::share_fs::{do_get_guest_path, do_get_host_path};
+use agent::Storage;
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use hypervisor::{device::DeviceManager, BlockConfig, DeviceConfig};
+use kata_types::mount::Mount;
+use nix::sys::stat::{self, SFlag};
+use std::fs;
+use tokio::sync::RwLock;
+
+pub(crate) struct BlockRootfs {
+    guest_path: String,
+    device_id: String,
+    mount: oci::Mount,
+    storage: Option<agent::Storage>,
+}
+
+impl BlockRootfs {
+    pub async fn new(
+        d: &RwLock<DeviceManager>,
+        sid: &str,
+        cid: &str,
+        dev_id: u64,
+        _bundle_path: &str,
+        rootfs: &Mount,
+    ) -> Result<Self> {
+        let container_path = do_get_guest_path(ROOTFS, cid, false, false);
+        let host_path = do_get_host_path(ROOTFS, sid, cid, false, false);
+        // Create rootfs dir on host to make sure mount point in guest exists, as readonly dir is
+        // shared to guest via virtiofs, and guest is unable to create rootfs dir.
+        fs::create_dir_all(&host_path)
+            .map_err(|e| anyhow!("failed to create rootfs dir {}: {:?}", host_path, e))?;
+
+        let block_device_config = &mut BlockConfig {
+            major: stat::major(dev_id) as i64,
+            minor: stat::minor(dev_id) as i64,
+            ..Default::default()
+        };
+
+        let device_id = d
+            .write()
```

Comment:
Ditto, only one `d.write()` may be ok?

Reply:
The reason is the same as above, just for fine-grained lock